**Федюкин Д.А. ИУ5-62б**

In [0]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
import numpy as np
import pandas as pd
import scipy
from sklearn.svm import  SVC
from sklearn.metrics import accuracy_score, roc_auc_score
import matplotlib.pyplot as plt
from sklearn import ensemble , model_selection, metrics 
from sklearn.model_selection import learning_curve

Задача классификации фамилий

#Prep

In [0]:
model = TfidfVectorizer(
    lowercase = False, 
    analyzer='char',
    use_idf = False,
    sublinear_tf=True,
    ngram_range=(1,4),
    max_features=50000)

In [0]:
data = pd.read_csv("linear_train.txt",sep=',', header=None)

In [0]:
#data[0]=data[0].str.lower()


In [0]:
data[0]=data[0].replace(r"[^а-яА-я0-9_]+","", regex=True)

In [0]:
#data[0]=data[0].apply(lambda x: x[-5:] if len(x)>5 else x)

In [0]:
data

0  1
0       Аалтонен  1
1            Аар  0
2          Аарон  0
3          ААРОН  0
4         Аарона  0
...          ... ..
101403      лкин  1
101404      лкой  0
101405       лок  0
101406    лочкой  0
101407    мкость  0

[101408 rows x 2 columns]

In [0]:
data.drop_duplicates(inplace=True, subset=0)

In [0]:
data_vectorized = model.fit_transform(data[0])

In [0]:
train_data, test_data, train_target, test_target = (
train_test_split(data_vectorized, data[1], test_size = 0.3, random_state=0))

#Random Forest

In [0]:
rfc = ensemble.RandomForestClassifier(n_estimators = 300, random_state=1)

In [0]:
rfc.fit(train_data,train_target)


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=300,
                       n_jobs=None, oob_score=False, random_state=1, verbose=0,
                       warm_start=False)

In [0]:
y1 = rfc.predict(test_data)

In [0]:
print(roc_auc_score(test_target,y1))

0.6755723386700261


ROC_AUC = 0.6755723386700261

# SVM

In [0]:
SVCCls = SVC()
SVCCls.fit(train_data, train_target)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [0]:
ANS = SVCCls.predict(test_data)

In [0]:
print(roc_auc_score(test_target, ANS))

0.698242465069693


ROC_AUC = 0.698242465069693

На моем наборе данных эффективнее оказался Машины опорных векторов